framework/src/Volo.Abp.Caching/Volo/Abp/Caching/DistributedCache.cs

```diff
@@ -52,23 +79,76 @@ public class DistributedCache<TCacheItem> : IDistributedCache<TCacheItem>
             _cacheOption = cacheOption.Value;
             Cache = cache;
             CancellationTokenProvider = cancellationTokenProvider;
-            Logger = NullLogger<DistributedCache<TCacheItem>>.Instance;
+            Logger = NullLogger<DistributedCache<TCacheItem, TCacheKey>>.Instance;
             Serializer = serializer;
             CurrentTenant = currentTenant;
 
             SyncSemaphore = new SemaphoreSlim(1, 1);
 
             SetDefaultOptions();
         }
+        protected virtual string NormalizeKey(TCacheKey key)
+        {            
+            Type type = key.GetType();
+            string keyValue = key.ToString();
+
+            // If complex type of object, override the key value with property concatenation
+            if (!typeof(IComparable).IsAssignableFrom(type) || type.IsPrimitive || type.IsValueType)
```

Comment:
We should not calculate the key value. We just call key.ToString() and use it as the key value to let the developer override ToString() and return a the value for a complex object.
We document it, so developers know what to do to be able to use a complex object as the cache key.